gui: grey out used address in address book

za-kk opened the pull request:
Implements issue #17174, to grey out used addresses in the address book when the wallet has the `avoid_reuse` flag set.

this commit brings the `IsUsedDestination` method into the `Wallet.h` interface. It is then called in `addresstablemodel.cpp` to determine whether the address has been used or not whilst setting the font colour

master
<img width="840" alt="Screenshot 2019-11-02 at 17 36 51" src="https://user-images.githubusercontent.com/31032215/68075122-897dec00-fd9b-11e9-95ac-4a8f36635cc9.png">

pr
<img width="836" alt="Screenshot 2019-11-02 at 17 33 41" src="https://user-images.githubusercontent.com/31032215/68075130-9e5a7f80-fd9b-11e9-93e3-6d687b982c34.png">

promag commented on src/interfaces/wallet.cpp:
```diff
@@ -202,6 +202,12 @@ class WalletImpl : public Wallet
         LOCK(m_wallet->cs_wallet);
         return m_wallet->ListLockedCoins(outputs);
     }
+    bool IsUsedDestination(const CTxDestination& dst) override
```
There was some discussion elsewhere which, IIRC, say you should use `isUsedDestination` instead.

za-kk replied:
yes, makes sense as it follows the same convention as the rest of the file. will update along with other nits

za-kk replied:
fixed

promag commented on src/interfaces/wallet.h:
```diff
@@ -132,6 +132,9 @@ class Wallet
     //! List locked coins.
     virtual void listLockedCoins(std::vector<COutPoint>& outputs) = 0;
 
+    // Whether this or any UTXO with the same CTxDestination has been spent.
```
The comment is not clear to me, what you mean by "this"?

za-kk replied:
this comment was carried over from the original `IsUsedDestination` method in the `wallet/wallet.h` file

promag commented on src/interfaces/wallet.cpp:
```diff
@@ -202,6 +202,12 @@ class WalletImpl : public Wallet
     }
+    bool IsUsedDestination(const CTxDestination& dst) override
+    {   
+        auto locked_chain = m_wallet->chain().lock();
```
Why are you locking this?

za-kk replied:
sorry, think i misunderstood what this line did, will remove

za-kk replied:
this has now been removed

promag commented on src/qt/addresstablemodel.cpp:
```diff
@@ -228,6 +228,13 @@ QVariant AddressTableModel::data(const QModelIndex &index, int role) const
         default: break;
         }
     }
+    else if (role == Qt::ForegroundRole) 
```
nit, here and below `if (...) {`.

za-kk replied:
fixed

promag commented on src/qt/addresstablemodel.cpp:
```diff
@@ -228,6 +228,13 @@ QVariant AddressTableModel::data(const QModelIndex &index, int role) const
     }
+    else if (role == Qt::ForegroundRole) 
+    {
+        if(walletModel->wallet().IsUsedDestination(DecodeDestination(rec->address.toStdString()))) 
```
nit, space after `if`.

za-kk replied:
fixed

luke-jr commented:
Addresses are used when they receive. Not when the UTXO created is later spent.

za-kk commented:
> Addresses are used when they receive. Not when the UTXO created is later spent.

Understood, I can update it to grey out just addresses that have received (Used) and not ones that have received and been spent if desired?

luke-jr commented:
> Understood, I can update it to grey out just addresses that have received (Used) and not ones that have received and been spent if desired?

Yes (it would be meaningless/erroneous to grey out addresses based on actions of the UTXOs created when they receive).

jonasschnelli commented:
Agree with @luke-jr. This is currently conceptual wrong.

MarcoFalke commented:
So if the existing used-flag is supposed to do what it does (indicate true when the coin sent to that destination has been spent), it could make sense to add a new flag (indicate true when a tx has been received that has the destination in one of its outputs).

luke-jr commented:
See also #15987, but I'm not sure it's the best approach to use here. If we're going to check *every address* (as this does), we should probably save a flag in the wallet/address book when we see an address used...

za-kk commented:
Revisiting this pr...

I have now re-implemented by saving a flag in the address book (`address_used`) when we see an address used. I took the same approach as the `IsUsedDestination` methods and therefor this will only work when the `avoid_reuse` flag is set on the wallet.

This essentially checks the outputs when a transaction is added to the wallet (`AddToWallet`) and if it belongs to us, sets a flag in the address book (`SetUsedAddressState`). It will then grey out any addresses with this flag set when viewing them in the 'Receiving Addresses' dialog box.

dannmat commented:
ack

luke-jr commented:
IMO this should work for all wallets, and when loading an old wallet (or one that has been loaded by an older version since the last use of a supporting version), should populate the flag as necessary.

za-kk commented:
> IMO this should work for all wallets, and when loading an old wallet (or one that has been loaded by an older version since the last use of a supporting version), should populate the flag as necessary.

@luke-jr I would be inclined to agree with you on this, it was only initially constrained to `avoid_reuse` wallets as `IsUsedDestination` was used in the first integration (although wrongly used due to confusion). So pivoting back to working for all wallets would make sense.

Would like to get others opinions on this before progressing with making the changes, thoughts @MarcoFalke @jonasschnelli @promag?

za-kk commented:
Rebased with master and squashed commits.

Thinking of picking this back up if it is still a desirable change/feature? Currently it's limited to watch only wallets but as @luke-jr mentioned, this may be desirable to all wallet types

jonasschnelli commented:
Concept ACK on how this works today. Slightly reviewed the code and it looks good.

Can you please review @luke-jr, @promag?

MarcoFalke commented:
Open-Close to re-run ci. See https://github.com/bitcoin/bitcoin/issues/15847#issuecomment-487591130

promag commented on src/wallet/wallet.cpp:
```diff
@@ -784,6 +784,27 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+{
+    AssertLockHeld(cs_wallet);
+    CTxDestination dst;
+    if(ExtractDestination(cout.scriptPubKey, dst)) {
```
Please update code format accordingly to developer notes.

za-kk replied:
Updated according to developer notes (added space after `if`), hopefully I haven't missed anything, Thanks!

jonatack replied:
If helpful, you can use Clang in the future to check for some things:

    Use Clang formatting to clean up a file:
    clang-format -i <file>

    To format the last commit with 0 lines of context, call the
    clang-format-diff.py script from root as follows:
    git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v

za-kk replied:
Thanks @jonatack, that will definitely be useful for me to use in the future!

promag commented on src/wallet/wallet.cpp:
```diff
@@ -784,6 +784,27 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+bool CWallet::IsUsedAddress(const CTxDestination& dst) const
+{
+    LOCK(cs_wallet);
+    return IsMine(dst) && GetDestData(dst, "address_used", nullptr);
```
Is it me or `IsMine` is irrelevant?

Either way call `GetDestData` since it's faster and most likely to fail.

za-kk replied:
Yes `IsMine` would be irrelevant in this case, have now removed to just return the output of `GetDestData` as suggested, Thanks!

promag reviewed (commented):
Maybe add a tooltip on the greyed rows?

za-kk commented:
> Maybe add a tooltip on the greyed rows?

Added tooltip as suggested, not sure on suitable wording so suggestions welcome (I went with "This address has been used" followed by the original tooltip in brackets (screenshot below).

![address-tooltip](https://user-images.githubusercontent.com/31032215/83955638-d69fd000-a84c-11ea-95bf-4d1d365b51e8.png)

jonatack commented:
Concept ACK, will review.

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -784,6 +784,27 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+            }
+        }
+    }
+}
```
perhaps simplify
```diff
 {
     AssertLockHeld(cs_wallet);
     CTxDestination dst;
-    if (ExtractDestination(cout.scriptPubKey, dst)) {
-        if (IsMine(dst)) {
-            if (used && !GetDestData(dst, "address_used", nullptr)) {
-                AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
-            } else if (!used && GetDestData(dst, "address_used", nullptr)) {
-                EraseDestData(batch, dst, "address_used");
-            }
+    if (ExtractDestination(cout.scriptPubKey, dst) && IsMine(dst)) {
+        if (used && !GetDestData(dst, "address_used", nullptr)) {
+            AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
+        } else if (!used && GetDestData(dst, "address_used", nullptr)) {
+            EraseDestData(batch, dst, "address_used");
         }
     }
 }
```

za-kk replied:
Yes makes sense, this has now been simplified as suggested

jonatack commented on src/qt/addresstablemodel.cpp:
```diff
@@ -228,6 +228,16 @@ QVariant AddressTableModel::data(const QModelIndex &index, int role) const
     }
+    else if (role == Qt::ForegroundRole) {
+        if (rec->type == AddressTableEntry::Receiving && walletModel->wallet().isUsedAddress(DecodeDestination(rec->address.toStdString()))) {
+            return QColor(Qt::lightGray);
```
Suggestion: perhaps `gray` to be more compatible with dark mode.

gray:
![Screenshot from 2020-06-07 10-47-18](https://user-images.githubusercontent.com/2415484/83964322-d1b63d00-a89b-11ea-8294-f365f60395ae.jpg)

light gray:
![Screenshot from 2020-06-07 10-41-39](https://user-images.githubusercontent.com/2415484/83964324-d4b12d80-a89b-11ea-86a7-39b2845ac950.jpg)

za-kk replied:
Thanks for testing with dark mode! have now updated the code to use `gray` instead of `lightGray` to accommodate both light and dark modes

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -784,6 +784,27 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+
+bool CWallet::IsUsedAddress(const CTxDestination& dst) const
+{
+    LOCK(cs_wallet);
```
Is this lock needed?

za-kk replied:
Probably not needed as `GetDestData` is just accessing the address book data. I have now removed the lock.

jonatack replied:
Sorry about that; it appears to be needed.
```
wallet/wallet.cpp:803:12: error: calling function 'GetDestData' requires holding mutex 'cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
    return GetDestData(dst, "address_used", nullptr);
           ^
1 error generated.
```

za-kk replied:
My bad, should have checked properly myself before removing! This has now been updated to include the lock again.

jonatack reviewed (commented):
Concept ACK -- nice!

Builds cleanly and works as advertised, including the tooltip.

za-kk commented:
Thanks for reviewing the code @jonatack! I have now made the changes suggested and squashed my commits.

jonatack commented:
Thanks for the color change. I only use the GUI in dark mode :). I don't think you needed to squash the two commits and `git diff 94cbeaa 481d3df` would have been cleaner and easier to re-review by not squashing. But wait until someone else reviews before re-touching.

promag commented:
I'm sorry to NACK current approach. You shouldn't be adding blocking calls in `AddressTableModel::data`, these will make the GUI unresponsive. I think you should add a `is_used` member to `AddressTableEntry` and keep it updated accordingly.

za-kk commented:
Thanks for reviewing @promag! makes sense to avoid locking up the GUI wherever possible, I will look into updating the code to follow the process you mentioned instead.

hebasto commented:
Concept ACK.

@za-kk Are you still working on this? Are you going to implement @promag's [suggestion](https://github.com/bitcoin/bitcoin/pull/17355#issuecomment-640480797)?

za-kk commented:
Thanks for the ACK @hebasto, yes I am still working on this and plan to implement the suggestions made by @promag soon 👍🏻

jonatack commented:
Hi @za-kk, let us know if you're having any issue with progress here :rocket:

jarolrod commented:
@za-kk  are you going to implement promags suggestion, or can I pick this up?

hebasto commented:
> @za-kk are you going to implement promags suggestion, or can I pick this up?

Probably, the better approach is to split this change into two PRs:
- introduce new `CWallet` member functions and _add tests_ for them (in this repo)
- GUI changes (in the gui repo)

za-kk commented:
Sorry, this PR has been quite stale from me over the last several months as I haven't had the chance to take a look yet. I do still plan on implementing the suggestions over the next week or so but always willing to accept help on this!

Thanks @hebasto, I was wondering what the preferred way of doing this would be now that we have the GUI repo. I will look into introducing the initial `CWallet` functions over the coming days

promag commented:
@hebasto I don't quite agree with that approach. The goal is to add a GUI change and that PR should be in gui repo. If that PR requires `CWallet` changes then I think they belong there. It might make sense to do what you suggest but not in this early stage - too much effort to sync both PR's, to have split discussions etc.

hebasto commented:
@promag 
> @hebasto I don't quite agree with that approach. The goal is to add a GUI change and that PR should be in gui repo. If that PR requires `CWallet` changes then I think they belong there. It might make sense to do what you suggest but not in this early stage - too much effort to sync both PR's, to have split discussions etc.

I'm fine with keeping all changes and discussion here.

New introduced functions still require test coverage (probably, in a follow up).

adamjonas commented on src/wallet/wallet.cpp:
```diff
@@ -842,6 +842,25 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
 
+void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, bool used)
+{
+    AssertLockHeld(cs_wallet);
```
I'm getting compiler warnings consistent with the CI:

```
wallet/wallet.cpp:847:5: warning: calling function 'AssertLockHeldInternal<AnnotatedMixin<std::__1::recursive_mutex> >' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
    AssertLockHeld(cs_wallet);
    ^
./sync.h:81:28: note: expanded from macro 'AssertLockHeld'
#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
                           ^
wallet/wallet.cpp:849:55: warning: calling function 'IsMine' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
    if (ExtractDestination(cout.scriptPubKey, dst) && IsMine(dst)) {
                                                      ^
wallet/wallet.cpp:850:22: warning: calling function 'GetDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
        if (used && !GetDestData(dst, "address_used", nullptr)) {
                     ^
wallet/wallet.cpp:851:13: warning: calling function 'AddDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
            AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
            ^
wallet/wallet.cpp:852:29: warning: calling function 'GetDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
        } else if (!used && GetDestData(dst, "address_used", nullptr)) {
                            ^
wallet/wallet.cpp:853:13: warning: calling function 'EraseDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
            EraseDestData(batch, dst, "address_used");
            ^
6 warnings generated.
```

za-kk replied:
Thanks for taking a look @adamjonas, this should now be resolved in latest commit

promag reviewed (commented):
I think should do the following:
 - start by adding a new `is_used` member  in `interfaces::WalletAddress` 
 - then in `getAddresses()` (src/wallet/interfaces.cpp) `is_used` should be set - you have the wallet lock there
 - add a `is_used` member to `AddressTableEntry`
 - copy `is_used` to entry in `AddressTablePriv::refreshAddressTable`
 - ensure `is_used` is updated accordingly in `AddressTablePriv::updateEntry`
 - expose in the model (like you do in this PR) but read data only from the record.

I hope this helps.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -842,6 +842,25 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+    CTxDestination dst;
+    if (ExtractDestination(cout.scriptPubKey, dst) && IsMine(dst)) {
+        if (used && !GetDestData(dst, "address_used", nullptr)) {
+            AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
```
Could be save txid instead of "p" and maybe reword the key to "first_txid" so that we could show the user that transaction.

za-kk replied:
This should be implemented in latest commit

promag commented on src/wallet/wallet.cpp:
```diff
@@ -842,6 +842,25 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
     return false;
 }
 
+void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, bool used)
```
`used` is always `true`, maybe drop it?

za-kk replied:
This has been resolved in the latest commit, thanks!

promag commented on src/wallet/wallet.cpp:
```diff
@@ -842,6 +842,25 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+bool CWallet::IsUsedAddress(const CTxDestination& dst) const
+{
+    AssertLockHeld(cs_wallet);
+    return GetDestData(dst, "address_used", nullptr);
```
An existing wallet will not have these set, ignore this fact? Or just add a release note saying it that a rescan fixes this?

za-kk replied:
Yes I think the best solution would be adding a release note as you suggested, I will look into doing this shortly

hebasto commented:
> Hi @za-kk, let us know if you're having any issue with progress here

za-kk commented:
Hi All, I finally had chance to implement the changes suggested by @promag. I notice that the test `wallet_avoidreuse.py` is currently failing in CI, I will look into resolving this.

hebasto commented:
> @za-kk Are you still working on this?

Any update?

hebasto commented:
@MarcoFalke @meshcollider 

Up for grabs?

za-kk commented:
Apologies for the delay on this PR, I do plan on picking it back up very shortly. 

Always willing to accept help/advice to move things forward @hebasto 👍🏻

CoinCornerZakk committed:
grey out used addresses in address book if avoid_reuse flag has been set on wallet

za-kk commented:
This should now be ready for review again, I have implemented the suggested changes made by @promag.

prayank23 commented:
Concept ACK

nit: grey if used once and red if more than 1 would have been better

Will test by tomorrow

jonatack commented on src/wallet/wallet.h:
```diff
@@ -455,6 +455,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     std::vector<OutputGroup> GroupOutputs(const std::vector<COutput>& outputs, const CoinSelectionParams& coin_sel_params, const CoinEligibilityFilter& filter, bool positive_only) const;
+    //! Whether this address has been used.
+    void SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, uint256 hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Here and in the definition in the .cpp file, `uint256 hash` ought to be passed by reference to const, as it is not a cheaply passed type (see usage elsewhere in the codebase). 
```suggestion
    void SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const

za-kk replied:
Thanks @jonatack, this should now be resolved in bb4790816d84f850367c99134ea6a16659fbb180

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -860,6 +860,22 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
     return false;
 }
 
+void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, uint256 hash)
```
```suggestion
void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, const uint256& hash)
```

za-kk replied:
Thanks @jonatack, this should now be resolved in bb4790816d84f850367c99134ea6a16659fbb180

jonatack reviewed (commented):
Thanks for updating! This is a bit of a drive-by review as the hour is late, but I was unable to build (with clang 11 on debian) after rebasing on current master at 7a49fdc :

```rake
wallet/wallet.cpp:878:15: error: out-of-line definition of 'SetUsedAddressState' does not match any declaration in 'CWallet'
void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, const uint256& hash)
              ^~~~~~~~~~~~~~~~~~~
./wallet/wallet.h:460:70: note: type of 3rd parameter of member declaration does not match definition ('uint256' vs 'const uint256 &')
    void SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, uint256 hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
                                                                     ^
wallet/wallet.cpp:883:9: error: use of undeclared identifier 'AddDestData'
        AddDestData(batch, dst, "first_txid", hash.ToString());
        ^
wallet/wallet.cpp:884:9: error: no matching function for call to object of type 'boost::signals2::signal<void (const CTxDestination &, const std::string &, bool, const std::string &, ChangeType)>' (aka 'signal<void (const variant<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown> &, const basic_string<char> &, bool, const basic_string<char> &, ChangeType)>')
        NotifyAddressBookChanged(this, dst, m_address_book[dst].GetLabel(), IsMine(dst) != ISMINE_NO, "receive", CT_UPDATED );
        ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/boost/signals2/detail/signal_template.hpp:720:19: note: candidate function not viable: requires 5 arguments, but 6 were provided
      result_type operator ()(BOOST_SIGNALS2_SIGNATURE_FULL_ARGS(BOOST_SIGNALS2_NUM_ARGS))
                  ^
/usr/include/boost/signals2/detail/signal_template.hpp:724:19: note: candidate function not viable: requires 5 arguments, but 6 were provided
      result_type operator ()(BOOST_SIGNALS2_SIGNATURE_FULL_ARGS(BOOST_SIGNALS2_NUM_ARGS)) const
                  ^
wallet/wallet.cpp:891:12: error: use of undeclared identifier 'GetDestData'
    return GetDestData(dst, "first_txid", nullptr);
           ^
4 errors generated.
```

CoinCornerZakk committed:
rework to avoid using AddDestData so it builds

za-kk commented:
Thanks for taking a look @jonatack! seems that some latest merges to master has resulted in changes to destdata (#21353)  which was causing the build issues. I have now refactored the code to deal with this (bb4790816d84f850367c99134ea6a16659fbb180)

prayank23 reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/17355/commits/bb4790816d84f850367c99134ea6a16659fbb180

prayank23 commented:
Compiled successfully on Linux (Pop!_OS). Works as expected. Addresses were greyed out once I sent bitcoin to first 2 addresses from the 'receiving addresses':

![image](https://user-images.githubusercontent.com/13405205/124507507-86969280-ddeb-11eb-8035-cf554a82b209.png)


Although if a user cares out privacy, should never open 'received addresses' for using because once a receiving address is created, it should be considered used. Best practice is to create a new address every time you need an address for receiving bitcoin.

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -875,6 +875,33 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+    CTxDestination dst;
+    if (ExtractDestination(cout.scriptPubKey, dst) && IsMine(dst)) {
+
+        m_address_book[dst].destdata.insert(std::make_pair("first_txid", hash.ToString()));
```
```suggestion
        m_address_book[dst].destdata.emplace(std::make_pair("first_txid", hash.ToString()));
```

jonatack commented on src/qt/addresstablemodel.cpp:
```diff
@@ -231,6 +234,16 @@ QVariant AddressTableModel::data(const QModelIndex &index, int role) const
         assert(false);
     }
+    else if (role == Qt::ForegroundRole) {
+        if (rec->is_used == true) {
```
```suggestion
        if (rec->is_used) {
```

jonatack commented on src/qt/addresstablemodel.cpp:
```diff
@@ -231,6 +234,16 @@ QVariant AddressTableModel::data(const QModelIndex &index, int role) const
+        }
+    }
+    else if (role == Qt::ToolTipRole) {
+        if (rec->is_used == true) {
```
```suggestion
        if (rec->is_used) {
```

jonatack replied:
(or for both, maybe simplify further)
```diff
@@ -234,15 +234,11 @@ QVariant AddressTableModel::data(const QModelIndex &index, int role) const
         } // no default case, so the compiler can warn about missing cases
         assert(false);
     }
-    else if (role == Qt::ForegroundRole) {
-        if (rec->is_used == true) {
-            return QColor(Qt::gray);
-        }
+    else if (role == Qt::ForegroundRole && rec->is_used) {
+        return QColor(Qt::gray);
     }
-    else if (role == Qt::ToolTipRole) {
-        if (rec->is_used == true) {
-            return tr("This address has been used (Right-click to edit address or label)");
-        }
+    else if (role == Qt::ToolTipRole && rec->is_used) {
+        return tr("This address has been used (Right-click to edit address or label)");
     }
     return QVariant();
```

jonatack commented on src/wallet/wallet.h:
```diff
@@ -456,6 +456,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     std::vector<OutputGroup> GroupOutputs(const std::vector<COutput>& outputs, const CoinSelectionParams& coin_sel_params, const CoinEligibilityFilter& filter, bool positive_only) const;
+    //! Whether this address has been used (has received).
+    void SetAddressReceivedState(WalletBatch& batch, const CTxOut& cout, const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
The first argument of `SetAddressReceivedState()` seems unused unless I'm missing something.
```suggestion
    void SetAddressReceivedState(const CTxOut& cout, const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -875,6 +875,33 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+        }
+    }
+    return false;
+}
```
style nit: clang format and also can simplify
```diff
@@ -890,16 +890,8 @@ bool CWallet::HasAddressReceived(const CTxDestination& dst) const
 {
 
     const std::string key{"first_txid"};
-    std::map<CTxDestination, CAddressBookData>::const_iterator i = m_address_book.find(dst);
-    if(i != m_address_book.end())
-    {
-        CAddressBookData::StringMap::const_iterator j = i->second.destdata.find(key);
-        if(j != i->second.destdata.end())
-        {
-            return true;
-        }
-    }
-    return false;
+    const std::map<CTxDestination, CAddressBookData>::const_iterator it = m_address_book.find(dst);
+    return it != m_address_book.end() && it->second.destdata.find(key) != it->second.destdata.end() ? true : false;
 }
```

jonatack reviewed (commented):
Some comments for your consideration. The pull doesn't necessarily need to be done in one commit depending on how you organise the changes and provided they are hygienic, but the "fix up to make it build" commit probably should be squashed into the relevant changes. Don't hesitate to run clang-format on your changes.

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -875,6 +875,33 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
+    AssertLockHeld(cs_wallet);
+    CTxDestination dst;
+    if (ExtractDestination(cout.scriptPubKey, dst) && IsMine(dst)) {
+
```
nit, omit extra line break here and also line 891 below

ryanofsky reviewed (commented):
Reviewing bb4790816d84f850367c99134ea6a16659fbb180. This seems pretty good but:

- It seems like `SetAddressReceivedState` should be called in `LoadToWallet` as well as `AddToWallet`, not just `AddToWallet` so old and new addresses are displayed correctly, not just ~old~new addresses.
- It seems like `SetAddressReceivedState` should be called regardless of `WALLET_FLAG_AVOID_REUSE` flag. The avoid reuse feature is for detecting addresses that we previously _spent from_  (and avoiding spending from them again). This feature is about detecting which addresses we previously _received to_, which is not directly related, and probably doesn't need a feature flag since it doesn't affect the wallet database storage.
- Would seem more efficient and simpler to a add new bool `CAddressBookData::received` member instead of adding a `"first_txid"` entry and transaction hash to the `CAddressBookData::destdata` map
- Could be nice to have some test coverage for this in https://github.com/bitcoin/bitcoin/blob/master/src/qt/test/wallettests.cpp

EDIT: corrected old/new above

prayank23 commented:
> It seems like SetAddressReceivedState should be called in LoadToWallet as well as AddToWallet, not just AddToWallet so old and new addresses are displayed correctly, not just old addresses.

I relaunched `bitcoin-qt` today and my old addresses which had received bitcoin, were marked [grey yesterday](https://github.com/bitcoin/bitcoin/pull/17355#issuecomment-874272237) are not greyed out now:

![image](https://user-images.githubusercontent.com/13405205/124614070-8355e200-de91-11eb-896e-061f44d0dcab.png)

> It seems like SetAddressReceivedState should be called regardless of WALLET_FLAG_AVOID_REUSE flag. The avoid reuse feature is for detecting addresses that we previously spent from (and avoiding spending from them again). This feature is about detecting which addresses we previously received to, which is not directly related, and probably doesn't need a feature flag since it doesn't affect the wallet database storage.

Agree

za-kk commented:
Thanks for the feedback everyone! Closing this PR for now until I have more time to come back and implement the changes, anyone else is welcome to pick it up in the meantime if desired 😀
